src/ln/channelmanager.rs

```diff
@@ -2121,63 +2143,82 @@ impl ChannelManager {
-										c if c == UPDATE|13 => 4, // incorrect_cltv_expiry
-										c if c == UPDATE|14 => 0, // expiry_too_soon
-										c if c == UPDATE|20 => 2, // channel_disabled
-										_ =>  {
-											// node sending unknown code
-											res = Some((Some(msgs::HTLCFailChannelUpdate::NodeFailure {
-												node_id: route_hop.pubkey,
-												is_permanent: true,
-											}), false));
-											return;
-										}
+								_c if _c & 0xff00 == PERM => {
+									match error_code & 0xff {
+										8 => onion_failure_log!("permanent_channel_failure", error_code),
+										9 => onion_failure_log!("required_channel_feature_missing", error_code),
+										10 => onion_failure_log!("unknown_next_peer", error_code),
+										_ => return_node_fail_update!(res, route_hop.pubkey, true),
+									}
+									assert!(next_route_hop_ix < route.hops.len());
+									res = Some((Some(msgs::HTLCFailChannelUpdate::ChannelClosed {
+										short_channel_id: route.hops[next_route_hop_ix].short_channel_id,
+										is_permanent: true,
+									}), true));
+									return;
+								},
+								_c if _c & 0xff00 == UPDATE => {
+									let offset = match error_code & 0xff{
+										7  => 0, // temporary_channel_failure
+										11 => 8, // amount_below_minimum
+										12 => 8, // fee_insufficient
+										13 => 4, // incorrect_cltv_expiry
+										14 => 0, // expiry_too_soon
+										20 => 2, // channel_disabled
+										_ => return_node_fail_update!(res, route_hop.pubkey, true),
 									};
-
 									if err_packet.failuremsg.len() >= offset + 2 {
 										let update_len = byte_utils::slice_to_be16(&err_packet.failuremsg[offset+2..offset+4]) as usize;
 										if err_packet.failuremsg.len() >= offset + 4 + update_len {
 											if let Ok(chan_update) = msgs::ChannelUpdate::read(&mut Cursor::new(&err_packet.failuremsg[offset + 4..offset + 4 + update_len])) {
+
 												// if channel_update should NOT have caused the failure:
 												// MAY treat the channel_update as invalid.
-												let is_chan_update_invalid = match error_code {
-													c if c == UPDATE|7 => { // temporary_channel_failure
+												let is_chan_update_invalid = match error_code & 0xff {
+													7 => {
+														onion_failure_log!("temporary_channel_failure", error_code);
 														false
 													},
-													c if c == UPDATE|11 => { // amount_below_minimum
-														let reported_htlc_msat = byte_utils::slice_to_be64(&err_packet.failuremsg[2..2+8]);
-														onion_failure_log!("amount_below_minimum", UPDATE|11, "htlc_msat", reported_htlc_msat);
-														incoming_htlc_msat > chan_update.contents.htlc_minimum_msat
+													11 => {
+														onion_failure_log!("amount_below_minimum", error_code, "htlc_msat", byte_utils::slice_to_be64(&err_packet.failuremsg[2..2+8]));
+														amt_to_forward > chan_update.contents.htlc_minimum_msat
```

Comment:
switch to amt_to_forwatd because it's error code for the outgoing link so incoming_htlc_msat wasn't accurate right ?

Reply:
yes. 
`.. from the processing node.` per spec